File: src/main/java/cz/muni/fi/pv242/service/UserService.java
package cz.muni.fi.pv242.service;

import java.util.List;

import cz.muni.fi.pv242.rest.model.BorrowingCreateDTO;
import cz.muni.fi.pv242.rest.model.BorrowingDTO;
import cz.muni.fi.pv242.rest.model.ReservationCreateDTO;
import cz.muni.fi.pv242.rest.model.ReservationDTO;
import cz.muni.fi.pv242.rest.model.UserAuthenticateDTO;
import cz.muni.fi.pv242.rest.model.UserCreateDTO;
import cz.muni.fi.pv242.rest.model.UserDTO;
import cz.muni.fi.pv242.rest.model.UserUpdateDTO;

/**
 * Created by honza on 5/18/16.
 */
public interface UserService {

	
    UserDTO createUser(UserCreateDTO user);
    UserDTO updateUser(UserUpdateDTO user);
    UserDTO getUserById(long id);
    UserDTO getUserByEmail(String email);
    void enableUser(long u);
    boolean disableUser(long u);
    boolean authenticate(String email, String password);
    List<UserDTO> getAllUsers();
    List<ReservationDTO> getReservationsOfUser(long userId);
    List<BorrowingDTO> getBorrowingsOfUser(long userId);
    ReservationDTO addReservationToUser(long id, ReservationCreateDTO reservation);
    BorrowingDTO addBorrowingToUser(long id, BorrowingCreateDTO borrowing, long bookId);
    void removeReservationOfUser(long id, long reservationId);
    void removeBorrowingOfUser(long id, long borrowingId);
}

File: src/main/webapp/app-services/user.service.js
(function () {
    'use strict';

    angular
        .module('app')
        .factory('UserService', UserService);

    UserService.$inject = ['$http'];
    function UserService($http) {
        var service = {};

        service.GetAll = GetAll;
        service.GetById = GetById;
        service.GetByEmail = GetByEmail;
        service.Create = Create;
        service.Update = Update;
        service.EnableUser = EnableUser;
        service.DisableUser = DisableUser;
        service.GetReservationsOfUser = GetReservationsOfUser;
        service.GetBorrowingsOfUser = GetBorrowingsOfUser;
        service.AddReservationToUser = AddReservationToUser;
        service.AddBorrowingToUser = AddBorrowingToUser;
        service.RemoveReservationOfUser = RemoveReservationOfUser;
        service.RemoveBorrowingOfUser = RemoveBorrowingOfUser;

        return service;

        function GetAll() {
            return $http.get('/library/rest/users').then(handleSuccess, handleError('Error getting all users'));
        }

        function GetById(id) {
            return $http.get('/library/rest/users/' + id).then(handleSuccess, handleError('Error getting user by id'));
        }

        function GetByEmail(email) {
            return $http.get('/library/rest/users/findByEmail?email=' + email).then(handleSuccess, handleError('Error getting user by username'));
        }

        function Create(user) {
            return $http.post('/library/rest/users/add', user).then(handleSuccess, handleError('User with ' + user.email + ' already exist!'));
        }

        function Update(user) {
            return $http.put('/library/rest/users/' + user.id, user).then(handleSuccess, handleError('Error updating user'));
        }

        function EnableUser(id) {
            return $http.post('/library/rest/users/' + id + '/enable').then(handleSuccess, handleError('Error enabling user'));
        }
        
        function DisableUser(id) {
            return $http.post('/library/rest/users/' + id + '/disable').then(handleSuccess, handleError('Error disabling user'));
        }
        
        function GetReservationsOfUser(id){
        	return $http.get('/library/rest/users/' + id + '/reservations').then(handleSuccess, handleError('Error getting reservations of user'));
        }
        
        function GetBorrowingsOfUser(id){
        	return $http.get('/library/rest/users/' + id + '/borrowings').then(handleSuccess, handleError('Error getting borrowings of user'));
        }
        
        function AddReservationToUser(id, reservation){
        	return $http.post('/library/rest/users/' + id + '/add/reservation', reservation)
        		.then(handleSuccess, handleError('Error adding reservation to user'));
        }
        
        function AddBorrowingToUser(id, borrowing, bookId){
        	return $http.post('/library/rest/users/' + id + '/remove/borrowing/' + bookId, borrowing)
        		.then(handleSuccess, handleError('Error adding borrowing to user'));
        }
        
        function RemoveReservationOfUser(id, reservationId){
        	return $http.delete('/library/rest/users/' + id + '/remove/reservation/' + reservationId)
    			.then(handleSuccess, handleError('Error removing reservation of user'));
        }
        
        function RemoveBorrowingOfUser(id, borrowingId){
        	return $http.delete('/library/rest/users/' + id + '/remove/borrowing/' + borrowingId)
    			.then(handleSuccess, handleError('Error removing borrowing of user'));
        }


        function handleSuccess(res) {
            return res.data;
        }

        function handleError(error) {
            return function () {
                return error;
            };
        }
    }

})();
